sycl/include/sycl/ext/oneapi/matrix/matrix-jit-use.hpp

```diff
@@ -0,0 +1,656 @@
+        Ptr, stride, __spv::MatrixLayout::ColumnMajor,
+        spv_scope_traits<Group>::value);
+    break;
+  case layout::packed_a:
+    res.spvm = __spirv_JointMatrixLoadINTEL<
+        T, NumRows, NumCols, spv_matrix_use_traits<Use>::value,
+        spv_matrix_layout_traits<layout::unused>::value>(
+        Ptr, stride, __spv::MatrixLayout::PackedA,
+        spv_scope_traits<Group>::value);
+    break;
+  case layout::packed_b:
+    res.spvm = __spirv_JointMatrixLoadINTEL<
+        T, NumRows, NumCols, spv_matrix_use_traits<Use>::value,
+        spv_matrix_layout_traits<layout::unused>::value>(
+        Ptr, stride, __spv::MatrixLayout::PackedB,
+        spv_scope_traits<Group>::value);
+    break;
+  }
+#else
+  (void)sg;
+  (void)res;
+  (void)src;
+  (void)stride;
+  (void)MemL;
+  throw runtime_error("joint matrix is not supported on host device.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif // __SYCL_DEVICE_ONLY__
+}
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols, use Use,
+          access::address_space Space>
+inline __SYCL_ALWAYS_INLINE void joint_matrix_store(
+    Group sg,
+    joint_matrix<T, NumRows, NumCols, Use, layout::unused, Group> &src,
+    multi_ptr<T, Space> res, size_t stride, layout MemL) {
+#ifdef __SYCL_DEVICE_ONLY__
+  T *Ptr = res.get();
+  switch (MemL) {
+  default:
+    assert(false && "Invalid Memory Layout!");
+  case layout::row_major:
+    __spirv_JointMatrixStoreINTEL<
+        T, NumRows, NumCols, spv_matrix_use_traits<Use>::value,
+        spv_matrix_layout_traits<layout::unused>::value>(
+        Ptr, src.spvm, stride, __spv::MatrixLayout::RowMajor,
+        spv_scope_traits<Group>::value);
+    break;
+  case layout::col_major:
+    __spirv_JointMatrixStoreINTEL<
+        T, NumRows, NumCols, spv_matrix_use_traits<Use>::value,
+        spv_matrix_layout_traits<layout::unused>::value>(
+        Ptr, src.spvm, stride, __spv::MatrixLayout::ColumnMajor,
+        spv_scope_traits<Group>::value);
+    break;
+  case layout::packed_a:
+    __spirv_JointMatrixStoreINTEL<
+        T, NumRows, NumCols, spv_matrix_use_traits<Use>::value,
+        spv_matrix_layout_traits<layout::unused>::value>(
+        Ptr, src.spvm, stride, __spv::MatrixLayout::PackedA,
+        spv_scope_traits<Group>::value);
+    break;
+  case layout::packed_b:
+    __spirv_JointMatrixStoreINTEL<
+        T, NumRows, NumCols, spv_matrix_use_traits<Use>::value,
+        spv_matrix_layout_traits<layout::unused>::value>(
+        Ptr, src.spvm, stride, __spv::MatrixLayout::PackedB,
+        spv_scope_traits<Group>::value);
+    break;
+  }
+#else
+  (void)sg;
+  (void)src;
+  (void)res;
+  (void)stride;
+  (void)MemL;
+  throw runtime_error("joint matrix is not supported on host device.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif // __SYCL_DEVICE_ONLY__
+}
+
+template <typename Group, typename T1, typename T2, typename T3, size_t M,
```

Comment:
In the CUDA backend the layouts need to be provided to `joint_matrix_mad` because they must be known at compile time by the builtins. In the CUDA backend these layouts are inferred from the `joint_matrix` layout template parameters: see https://github.com/intel/llvm/blob/a8810552225b4cdb479650011766e9de8f5474b6/sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp#L825.

So if we want interfaces that match in both backends we need to add them here too, in the same way. For the user it should appear no different due to CTAD.

Reply:
seems strange, if we want support intel/cuda backend, why we don't need to provide matrix_load without memL.